File: javascript150/138.ts
class _Node {
    val: number
    next: _Node | null
    random: _Node | null

    constructor(val?: number, next?: _Node, random?: _Node) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
        this.random = (random===undefined ? null : random)
    }
}


let visited = new Map()
function copyRandomList(head: _Node | null): _Node | null {
    if (!head) return null
    if (visited.has(head)) return visited.get(head)

    let newHead = new _Node(head?.val)
    visited.set(head, newHead)

    newHead.next = head!.next?copyRandomList(head!.next): null
    newHead.random = head!.random?copyRandomList(head!.random): null
    return newHead
};

File: RandomPractice/138.py
from typing import Optional


class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        newHead = Node(head.val)
        
        oldCurrent = head
        newCurrent = newHead

        oldDict = {}
        newDict = {}

        counter = 0
        oldDict[oldCurrent] = counter
        newDict[counter] = newCurrent

        while oldCurrent.next != None:
            counter += 1

            oldCurrent = oldCurrent.next
            newCurrent.next = Node(oldCurrent.val)
            newCurrent = newCurrent.next

            oldDict[oldCurrent] = counter
            newDict[counter] = newCurrent

        oldCurrent = head
        newCurrent = newHead

        while oldCurrent != None:
            if oldCurrent.random != None:    
                index = oldDict[oldCurrent.random]
                newCurrent.random = newDict[index]

            oldCurrent = oldCurrent.next
            newCurrent = newCurrent.next

        return newHead